builtin/providers/aws/auth_helpers.go

```diff
@@ -0,0 +1,170 @@
+package aws
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	awsCredentials "github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/iam"
+)
+
+const DefaultMetadataApiHost = "http://169.254.169.254:80"
+
+type IamInfoFromMetadata struct {
+	Code               string
+	LastUpdated        time.Time
+	InstanceProfileArn string
+	InstanceProfileId  string
+}
+
+func getAccountId(iamconn *iam.IAM, authProviderName string) (string, error) {
+	// If we have creds from instance profile, we can use metadata API
+	if authProviderName == "EC2RoleProvider" {
+		log.Println("[DEBUG] Trying to get account ID via AWS Metadata API")
+		url := buildMetadataUrl("/latest/meta-data/iam/info")
+		c := http.Client{
+			Timeout: 3 * time.Second,
+		}
+
+		resp, err := c.Get(url)
+		if err != nil {
+			return "", fmt.Errorf("Failed getting account ID via EC2 metadata API: %s", err)
+		}
+
+		if !responseContainsEC2Header(resp) {
+			return "", fmt.Errorf("Expected EC2 metadata API at %q, no EC2 headers found.", url)
+		}
+
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return "", fmt.Errorf("Failed reading EC2 metadata API response from %q: %s", url, err)
+		}
+		info := IamInfoFromMetadata{}
+		err = json.Unmarshal(body, &info)
+		if err != nil {
+			return "", fmt.Errorf("Failed parsing EC2 metadata API response from %q: %s", url, err)
+		}
+
+		return parseAccountIdFromArn(info.InstanceProfileArn)
+	}
+
+	// Then try IAM GetUser
+	log.Println("[DEBUG] Trying to get account ID via iam:GetUser")
```

Comment:
Ah nevermind, I was thinking we can exclude `AssumeRoleProvider` from the `GetUser` path, but it's more cumbersome than helpful